Mage/src/main/java/mage/cards/ExpansionSet.java

```diff
@@ -295,6 +292,32 @@ public void shuffleCollator() {
                 .collect(Collectors.toList());
     }
 
+    public void openBoosterBox() {
+        // nothing to do if not using collator
+        if (boosterCollator == null) {
+            return;
+        }
+        synchronized (boosterBox) {
```

Comment:
What's wrong. There are only one instance per `Set` in whole server, so it must be unchangeable (sets and card infos are like database). Booster box logic must be moved to draft implementation. Same as boosters does. So `ExpansionSet.openBoostersBox` must return booster box object with all needed data and booster gen/choose logic.

![shot_210922_115654](https://user-images.githubusercontent.com/8344157/134304982-7f8f8d46-f2bf-490d-9c90-3487f89f7aa2.png)
![shot_210922_115703](https://user-images.githubusercontent.com/8344157/134304998-af4dde27-3ebf-4e2e-91d1-acf88e6d344c.png)

Reply:
If my code is the wrong approach, then the existing collation code by @theelk801, which you approved on Feb 12 with ["Code looks fine"](https://github.com/magefree/mage/pull/7529#pullrequestreview-589608235) is wrong too.

Each `ExpansionSet` that uses collation has a single instance of `BoosterCollator`. `BoosterCollator` contains fields of various classes derived from `mage.collation.Rotator`, which is mutable:

![image](https://user-images.githubusercontent.com/9252545/134322446-ff62d1bf-d7b8-47b6-bf2f-8f0e37656834.png)

Every time a booster is created via `createBoosterUsingCollator`, the `position` fields of all the `Rotator`-derived objects inside the `BoosterCollator` are repeatedly mutated (and prior to this PR, this was done without any synchronization or locking!) 

If you like, I can rewrite `BoosterCollator` to be stateless (which means that each booster created via `BoosterCollator` will be completely independent of previously-created boosters, just like boosters created using the old method are) I'm afraid that rewriting the draft and sealed tournament implementations like you suggest will be more complicated and my Java programming ability might not be up to it.

Reply:
> Each ExpansionSet that uses collation has a single instance of BoosterCollator

Interesting finds. Each set have one `BoosterCollator`... `shuffle()` method calls only one time on set's init, but `makeBooster()` calls multiple times from multiple places and it uses shared `RarityConfiguration`/`Rotater`.

Yes, collator logic must be reworked for independent booster generation between server's tables/games.

> will be completely independent of previously-created boosters

Nope, it must depends on previous booster like paper does (#8179). 

How it can be implemented (well, all code and logic are same except origin of the objects):
* each draft uses booster boxes instead sets (draft creates a boxes list on init from a settings/sets list);
* boxes can use lazy init (e.g. create one or all boosters on first request, on no more boosters in the box, etc -- it depends on booster generation performance for richman/cubes -- need research for best approach);
* all boosters logic must work with booster boxes;
* each set can generate different booster boxes (with different collator and other settings);
* booster box must implements collator/random logic;

P.S. If no one takes that task, then I'll do it at the end of the week in that PR.

Reply:
>it must depends on previous booster like paper does

The entire point of #8179 is that we *don't want* boosters to depend on previous boosters in a way that gives strategic information to players that isn't accurate to paper Magic--e.g. having N copies of of the same card in your Sealed pool shouldn't allow you to deduce that your tournament opponents can't have any copies of that same card.

Reply:
Hmm. I must have misunderstood the purpose of these booster boxes.

So:
* real life collator uses multiple shuffles between boxes (e.g. some boosters from one box, another boosters from another box);
* old xmage collator uses one shuffle before start (e.g. all boosters from one box);
* new collator uses multiple shuffles between boxes (e.g. some boosters from one box, another boosters from another box). Like real life boxes works.

Is that right? Or you want to make more random (more random than in real life)?

Reply:
That's pretty much right. Ideally I would like boosters to come from the same box *as long as that doesn't leak unwarranted information to players* (i.e. within each individual player's sealed pool, or at the start of each draft round) but from different boxes when using the same box would give away information that impacts strategic decisions.

How accurate that is to real life depends largely on the size of the real-life event we're talking about (a FNM with a single draft table, or a GP with hundreds of players?) and how the event is managed (do they allow players to bring their own boosters, or do they only use newly-opened boxes?)

I believe it is **much, much better to err on the side of being more random** than real life (make each booster completely independent of each other booster, like xmage's original booster creation method) than to err on the side of not being random enough and give players information they shouldn't have about their opponents' sealed pools or their potential future draft picks.

Reply:
Mmm... Call `shuffle()` before each booster can "fix" all the problems and no needs in boxes at all?

Reply:
That would certainly be "more random than real life" but each individual booster that we create will be accurate to real-life boosters.

If we do it this way, then we should create a new BoosterCollator object and shuffle it each time we create a booster, rather than have it be a permanent field of `ExpansionSet`. In other words `ExpansionSet` would have this method:

```
protected boosterCollator getCollator() {
    return null;
}
```

and each set that has a collator will override this method with one that returns a new collator instance.

Also, I think we need to fix how `Rotator` is constructed. Right now it is initializing `items` to a view of the vararg passed to its constructor, and later shuffling `items` in place. I might be wrong but I don't think that's safe or correct, and we need to create a new list instead:

```
public Rotater(boolean keepOrder, T... items) {
    this.items = new ArrayList<>(Arrays.asList(items));
    this.keepOrder = keepOrder;
}
```